src/simplify_algebra.cpp

```diff
@@ -346,20 +555,110 @@ struct find_add_convs
+    return !(dots < 2 and convs < 2);
+}
+
+struct find_conv_dot_horiz_fusion
+{
+    auto matcher() const { return horiz_conv_dot(); }
+
+    static bool reaches(instruction_ref ins, const std::vector<instruction_ref>& args)
+    {
+        auto it = std::find_first_of(
+            ins->outputs().begin(), ins->outputs().end(), args.begin(), args.end());
+        if(it != ins->outputs().end())
+            return true;
+        return std::any_of(
+            ins->outputs().begin(), ins->outputs().end(), [&](auto i) { return reaches(i, args); });
+    }
+
+    void apply(program& p, const match::matcher_result& r) const
+    {
+        auto ins = r.result;
+
+        auto pred = [](auto i, auto j) {
+            if(i->get_operator() != j->get_operator())
+                return false;
+            if(not contains({"dot", "convolution"}, i->name()))
+                return true;
+            auto x = i->inputs()[1]->get_shape().lens();
+            auto y = j->inputs()[1]->get_shape().lens();
+            if(x.size() != y.size())
+                return false;
+            // Check that non-axises match
+            int axis = 1;
+            if(i->name() == "dot")
+            {
+                axis = x.size() - 1;
+            }
+            return axis_equal(x, y, axis);
+        };
+
+        auto each = [&](auto start, auto last) {
+            if(std::distance(start, last) < 2)
+                return;
+            auto&& name = (*start)->name();
+            if(not contains({"dot", "convolution"}, name))
+                return;
+            auto input = (*start)->inputs().front();
+            std::vector<instruction_ref> args;
+            std::transform(
+                start, last, std::back_inserter(args), [&](auto x) { return x->inputs().at(1); });
+            int axis        = 1;
+            int concat_axis = 0;
+            if(name == "dot")
+            {
+                axis        = int(args.front()->get_shape().lens().size() - 1);
+                concat_axis = axis;
+            }
+
+            for(auto arg : args)
+                p.move_instructions(arg, input);
+            // TODO: Check if axises match
+            auto concat = p.insert_instruction(input, op::concat{concat_axis}, args);
+            auto fused =
+                p.insert_instruction(std::next(input), (*start)->get_operator(), input, concat);
+            int64_t offset = 0;
+            for(auto arg : range(start, last))
+            {
+                int64_t len = arg->get_shape().lens()[axis];
+                p.replace_instruction(arg, op::slice{{axis}, {offset}, {offset + len}}, fused);
+                offset += len;
+            }
+        };
+
+        auto outputs = ins->outputs();
+        group_by(outputs.begin(), outputs.end(), each, pred);
+    }
+};
+
 void simplify_algebra::apply(program& p) const
 {
     // Run simplifications multiple times
-    for(int i = 0; i < 4; i++)
+    for(int i = 0; i < 8; i++)
```

Comment:
should we set the number of iterations to an environment variable? I am assuming it was changed from 4 to 8 because with 4 iterations the program wasn't fully simplified.

Reply:
I dont know if we need an env variable, I am not sure why we would need to change it. Ideally, I would like to change this to stop loops when there is no more matches.